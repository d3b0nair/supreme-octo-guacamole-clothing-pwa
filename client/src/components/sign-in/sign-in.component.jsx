import React, { useState, useEffect } from "react";
import { connect } from "react-redux";

import FormInput from "../form-input/form-input.component";
import CustomButton from "../custom-button/custom-button.component";

import {
  googleSignInStart,
  emailSignInStart,
} from "../../redux/user/user.actions";

import { createStructuredSelector } from "reselect";
import { selectUserError } from "../../redux/user/user.selector";

import {
  SignInContainer,
  TitleContainer,
  ButtonsBarContainer,
} from "./sign-in.styles.jsx";

import { TextMsg, TextContainer } from "../sign-up/sign-up.styles.jsx";

const ErrorMsgOnCredentialsDontMatch = ({ credentialsDontMatch, userError }) =>
  credentialsDontMatch ? (
    <TextContainer>
      <TextMsg>
        {userError.code === "auth/too-many-requests"
          ? " Access to this account has been temporarily disabled due to many failed login attempts. You can immediately restore it by resetting your password or you can try again later."
          : "‚ùå Password/Email doesn't match"}
      </TextMsg>
    </TextContainer>
  ) : null;

export function SignIn({ googleSignInStart, emailSignInStart, userError }) {
  const [credentialsDontMatch, setCredentialsDontMatch] = useState(null);
  const [credentials, setCredentials] = useState({ email: "", password: "" });
  const { email, password } = credentials;
  useEffect(() => {
    if (userError) {
      setCredentialsDontMatch(true);
    }
  }, [userError]);
  const handleSubmit = async (e) => {
    e.preventDefault();
    emailSignInStart(email, password);
  };

  const handleChange = (e) => {
    const { value, name } = e.target;
    setCredentials({ ...credentials, [name]: value });
  };

  return (
    <SignInContainer>
      <TitleContainer>I already have an account</TitleContainer>
      <span>Sign in with your email and password</span>
      <form onSubmit={handleSubmit}>
        <FormInput
          label="E-Mail"
          name="email"
          type="email"
          value={email}
          handleChange={handleChange}
          required
        />
        <FormInput
          label="Password"
          name="password"
          type="password"
          value={password}
          handleChange={handleChange}
          required
        />
        <ErrorMsgOnCredentialsDontMatch
          credentialsDontMatch={credentialsDontMatch}
          userError={userError}
        />

        <ButtonsBarContainer>
          <CustomButton type="submit">Sign in</CustomButton>
          <CustomButton
            data-testid="CustomButtonForGoogleSignInStart"
            type="button"
            onClick={googleSignInStart}
            isGoogleSignIn
          >
            Sign in with Google
          </CustomButton>
        </ButtonsBarContainer>
      </form>
    </SignInContainer>
  );
}
const mapStateToProps = createStructuredSelector({
  userError: selectUserError,
});
const mapDispatchToProps = (dispatch) => ({
  googleSignInStart: () => dispatch(googleSignInStart()),
  emailSignInStart: (email, password) =>
    dispatch(emailSignInStart({ email, password })),
});

export default connect(mapStateToProps, mapDispatchToProps)(SignIn);
