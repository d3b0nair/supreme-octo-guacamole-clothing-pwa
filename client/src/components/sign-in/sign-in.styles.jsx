import styled from "styled-components";

export const SignInContainer = styled.div`
  width: 320px;
  display: flex;
  flex-direction: column;
  @media screen and (max-width: 800px) {
    width: 100%;
    margin: auto;
  }
`;

export const TitleContainer = styled.h2`
  margin: 10px 0;
`;

export const ButtonsBarContainer = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-template-rows: 1fr;
  gap: 10px;
  justify-content: space-between;
  @media screen and (max-width: 800px) {
    flex-direction: column;
    margin-bottom: 40px;
  }
`;
