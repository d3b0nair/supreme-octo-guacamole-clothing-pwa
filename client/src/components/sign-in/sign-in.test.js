import React from "react";
import {
  render,
  fireEvent,
  waitFor,
  screen,
  cleanup,
} from "@testing-library/react";
import { SignIn } from "./sign-in.component";

describe("SignIn component", () => {
const mockGoogleSignInStart = jest.fn();
  const mockProps = (otherProps) => ({
    ...otherProps,
    googleSignInStart: mockGoogleSignInStart,
  });
  const component = (mockProps) => <SignIn {...mockProps} />;

  afterEach(cleanup);

  it("should mount", () => {
    const { container } = render(component(mockProps()));
    expect(container).toMatchSnapshot();
  });

  it("should call mockGoogleSignInStart when button clicked", async () => {
    render(component(mockProps()));
    await waitFor(() => fireEvent.click(screen.getByTestId("CustomButtonForGoogleSignInStart")));
    expect(mockGoogleSignInStart).toHaveBeenCalled();
  });
});
