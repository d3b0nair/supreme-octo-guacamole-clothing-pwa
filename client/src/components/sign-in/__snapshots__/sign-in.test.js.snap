// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn component should mount 1`] = `
<div>
  <div
    class="sc-eCImPb kiJsgT"
  >
    <h2
      class="sc-jRQBWg hRSpAi"
    >
      I already have an account
    </h2>
    <span>
      Sign in with your email and password
    </span>
    <form>
      <div
        class="sc-bdvvtL hMBGla"
      >
        <input
          class="sc-gsDKAQ hVBime"
          data-testid="FormInputContainer"
          name="email"
          required=""
          type="email"
        />
        <label
          class="sc-dkPtRN jpMQjL  form-input-label"
          data-testid="FormInputLabel"
        >
          E-Mail
        </label>
      </div>
      <div
        class="sc-bdvvtL hMBGla"
      >
        <input
          class="sc-gsDKAQ hVBime"
          data-testid="FormInputContainer"
          name="password"
          required=""
          type="password"
        />
        <label
          class="sc-dkPtRN jpMQjL  form-input-label"
          data-testid="FormInputLabel"
        >
          Password
        </label>
      </div>
      <div
        class="sc-gKclnd idUVbs"
      >
        <button
          class="sc-hKwDye hzDzeB"
          type="submit"
        >
          Sign in
        </button>
        <button
          class="sc-hKwDye hPNorZ"
          data-testid="CustomButtonForGoogleSignInStart"
          type="button"
        >
          Sign in with Google
        </button>
      </div>
    </form>
  </div>
</div>
`;
